A monolithic PR

tesfaldet opened the pull request:
1. Compatibility with hydra experiment overrides.
2. Quantized step for distributions now supported.
3. Sweep parameterization now more aligned with Hydra's standardized approach.
4. Sweep overrides now support RangeSweep properly via quantized uniform.
5. Wandb run id not passed through task function anymore, instead the task function is executed within a `wandb.init` context; the task function can access the run using `wandb.run` or can just directly use wandb functions like `wandb.log` without having to run `wandb.init`.
6. Also included support for the submitit launcher plugin to allow launching multiple agents in parallel.
7. A bunch of revolting monkeypatching on wandb's packages to allow for saving the `wandb` sweep directory to the hydra sweep directory. Even more monkeypatching to change the wandb sweep directory name to `.wandb` from `wandb` to follow the hydra convention of using hidden folders for saving. Hopefully future versions of wandb's client will expose some of the attributes necessary to switch the wandb directory more easily.
8. Plenty of bug fixes, code typing improvements, comments, code clarification, etc.
9. Moved main sweeper code to _impl.py so that it doesn't get accidentally loaded along with all its monkeypatching when the user doesn't want to use the sweeper.
10. Some pre-commit hook modifications.
11. Returning results and statuses from task function to launcher since wandb.agent doesn't expose results or anything, which is annoying. Relatedly, I improved task function error handling and propagation to the launcher so that the launcher is aware of job fails (the launcher always receives job success from the wrapped task function since the agent doesn't return anything, so instead the results are introspected from the task function and propagated upwards).
12. Added agent budget functionality for launching batches of agents until budget reached. This is meant for cluster usage, in which case `count` should be set to 1 so that each agent only runs its task function once, as recommended by the [wandb docs](https://docs.wandb.ai/guides/sweeps/faq#how-should-i-run-sweeps-on-slurm).
13. Can now pass in tags (as a list) and notes via the wandb_sweep_config config yaml.
14. Improved error handling for agent failures such as sweep not existing anymore, job failures unrelated to the agent, and run failures within an agent.
15. Pre-emption with `run.mark_preemption()` in conjunction with the submitit launcher's `checkpoint` capabilities. Not tested with RQ, Joblib, and Ray hydra launcher plugins.
16. Support for early termination, whether through hyperband or manually stopping/killing sweeps or manually stopping individual runs.
17. Grid search method now supported.

Remaining TODOs:

- [x] Test out resuming from a pre-empted job via submitit's `checkpoint` function and wandb's `mark_preemption` function. This will require having to override the `BaseSubmititLauncher`'s `checkpoint` function with your own that will call `wandb.mark_preemption()` and set `config.hydra.sweep.wandb_sweep_config.resume = True` before resubmitting the job to SLURM. Then when the job is running again, the new wandb agent will know to resume from a preempted state and will receive the same params as it did before from the cloud controller.
- [x] Implement max failure rate functionality so that the sweeper doesn't continue to launch more jobs when a certain percentage of the previously launched batch have failed.
- [x] Implement functionality for periodically checking the status of the wandb cloud (or local) sweep controller to avoid the situation of the launcher launching a batch of new jobs after the sweep was killed via the web interface
- [x] Implement budgeting functionality in a cluster environment to allow for launching batches of agents at a time until a max budget is reached. This is preferred in a cluster environment instead of using the `count` parameter since it is good-practice for a single cluster (SLURM) job to only run once and return its status code (job success or fail), i.e., [each job needs to correspond to a single agent with `count = 1`](https://docs.wandb.ai/guides/sweeps/faq#how-should-i-run-sweeps-on-slurm). EDIT: implemented, check example sweep [here](https://wandb.ai/tesfaldet/hydra_plugin_test_project_2/sweeps/tfj8bdfb?workspace=user-tesfaldet)
- [ ] ~~[Switching to lazy %-style of logging](https://docs.python.org/3/howto/logging.html#optimization) as pylint complains otherwise~~
- [x] Support for 'grid' sweep method. If user provides 'grid' as the search method, then we shouldn't be using SUPPORTED_DISTRIBUTIONS & "distribution" and instead return "value" or "values" if the user provides a single value or a list of values. This means we need to verify that "distribution" isn't provided by the user in any of the params to be sweeped through.
- [x] Test hyperband early termination

P.S. branch name has nothing to do with what was actually done.

tesfaldet committed:
Compatibility with hydra experiment overrides. Quantized step for distributions now supported. Sweep parameterization now more aligned with Hydra's standardized approach. Some bug fixes. Sweep overrides now support RangeSweep properly via quantized uniform. Wandb run now directly passed to task function instead of run id to avoid unnecessary init overhead.

tesfaldet committed:
Added some useful logs. Added some comments. Small changes.

tesfaldet commented:
I'll need to update the example to show how to use the updated parameterization style. Also, some unit tests will be necessary in the future so for now I'm going to paste this config as reference...

```yaml
# @package _global_
defaults:
  - override /hydra/sweeper: wandb

hydra:
  sweeper:
    wandb_sweep_config:
      name: test
      metric:
        name: best_avg_val_err
        goal: minimize
      method: bayes
      num_agents: 1
      entity: test_entity
      project: test_project
      # early_terminate:
      #   type: hyperband
      #   max_iter: 27  #  the maximum number of iterations.
      #   s: 2  # the total number of brackets
      #   eta: 3  # the bracket multiplier schedule
      count: 1  # recommended when running in SLURM environment. Each agent only runs one training job then exits.
      # sweep_id:  # used for resuming from a previous sweep
    params:
      # parameters are described using dot notation but will be accessible within the task function in typical OmegaConf
      # fashion, e.g., min_ca_iter = cfg.experiment.iter.train.ca.min
      experiment.iter.train.ca.min:
        distribution: log_uniform
        min: 16
        max: 18
      experiment.iter.train.ca.max:
        # should default to categorical distribution
        - 32
        - 64
        - 96
      experiment.pool_size:
        # should default to constant distribution
        - 256
      experiment.batch_size.train: 64  # should default to constant distribution
      experiment.batch_size.val: [16, 32]  # should default to categorical distribution
      model.cell_hidden_chns:
        # can also explicitly request a categorical distribution
        distribution: categorical
        values:
          - 8
          - 12
          - 16
          - 32
          - 64
          - 128
      model.embed_dim:
        # can also explicitly request a constant distribution
        distribution: constant
        value: 32
      model.depth:
        distribution: int_uniform
        min: 1.5  # should fail here since integers are required
        max: 2.0  # will fail here since integers are required but perhaps I could cast to int when int(floor(x)) == int(x)
      model.heads:
        # should have q=1 by default
        distribution: q_uniform
        min: 1
        max: 8
      model.head_dim:
        distribution: normal
        mu: 32
        sigma: 8
      model.mlp_dim:
        # should have mu=0 and sigma=1 by default
        distribution: q_log_normal
        q: 2
```

tesfaldet committed:
Monkeypatching some wandb functions to allow for using wandb's code save when in a hydra cwd that's diff than the code's directory. Being more explicit with function arg types.

tesfaldet commented:
<img width="923" alt="Screen Shot 2022-02-26 at 22 56 35" src="https://user-images.githubusercontent.com/5014902/155867617-72093e3c-cd04-4e59-ba91-80451b8b1632.png">

Need to fix `Command`. For some reason there's a bunch of junk after the program path (`masked_autoencoding/train.py`)

For reference, it should look something like this:
![Screen Shot 2022-02-27 at 10 57 09](https://user-images.githubusercontent.com/5014902/155889709-1ec38fbd-d361-4bc0-90fa-50139f1251b6.png)

tesfaldet commented:
Also need to deal with an issue where `wandb.sweep` creates an empty `wandb` directory in the original cwd (the code's working directory). This can be an issue when the hydra working directory is set to be different than the code's working directory, which is where `wandb.agent` and `wandb.init` save their results.

EDIT: fixed

tesfaldet committed:
Can now pass in tags (as list) and notes via wandb_sweep_config yaml

tesfaldet committed:
Logging to .wandb instead of wandb, had to resort to a lot of monkeypatching but it was worth it. Some code cleanup and clarification.

tesfaldet committed:
Merge remote-tracking branch 'origin/main' into run_id. Removed pre-emption logic for now.

tesfaldet commented:
![Screen Shot 2022-02-27 at 21 22 38](https://user-images.githubusercontent.com/5014902/155913746-d8b1a455-43ac-4f9f-83b5-9ce4135bfff5.png)
The weird command issue doesn't seem to happen when not using submitit. 

(the captain-pool repo is being shown since I'm now testing this using my fork of this repo with `origin` set to this repo's remote instead of my fork's. By default, wandb searches for the `origin` remote when saving git-related info)

tesfaldet committed:
Updated example. Fixed some comments. Making sure to use abs path for creating and using sweep directory. Updated README but still needs an update in the range section of categorical parameters since now it's using wandb's q_uniform.

tesfaldet commented:
An example sweep can be viewed [here](https://wandb.ai/tesfaldet/hydra_plugin_test_project/sweeps/sqzgdhhr?workspace=user-tesfaldet). So far I've tested this with the basic launcher and with the submitit launcher :) Hope you like the contributions!

tesfaldet commented:
Perhaps an improvement can be to allow the user to request for wandb to pick the run names instead of using hydra.sweep.subdir. Hydra.sweep.subdir is currently used for a run's name since it keeps the naming consistent between the names of the folders where each run's results are stored and the run's name. Hydra.sweep.dir is used for a run's group name for similar reasons.

tesfaldet commented:
There are some TODOs in there that I'm hoping to tackle in the coming days.


- [ ] Test out resuming from a pre-empted job via submitit's `checkpoint` function and wandb's `mark_preemption` function. This will require having to override the `BaseSubmititLauncher`'s `checkpoint` function with your own that will call `wandb.mark_preemption()` and set `config.hydra.sweep.wandb_sweep_config.resume = True` before resubmitting the job to SLURM. Then when the job is running again, the new wandb agent will know to resume from a preempted state and will receive the same params as it did before from the cloud controller.
- [x] Implement budgeting functionality in a cluster environment to allow for launching batches of agents at a time until a max budget is reached. This is preferred in a cluster environment instead of using the `count` parameter since it is good-practice for a single cluster (SLURM) job to only run once and return its status code (job success or fail), i.e., [each job needs to correspond to a single agent with `count = 1`](https://docs.wandb.ai/guides/sweeps/faq#how-should-i-run-sweeps-on-slurm). EDIT: implemented, check example sweep [here](https://wandb.ai/tesfaldet/hydra_plugin_test_project_2/sweeps/tfj8bdfb?workspace=user-tesfaldet)
- [ ] [Switching to lazy %-style of logging](https://docs.python.org/3/howto/logging.html#optimization) as pylint complains otherwise
- [ ] Support for 'grid' sweep method. If user provides 'grid' as the search method, then we shouldn't be using SUPPORTED_DISTRIBUTIONS & "distribution" and instead return "value" or "values" if the user provides a single value or a list of values. This means we need to verify that "distribution" isn't provided by the user in any of the params to be sweeped through.

These can be in separate PRs however. This PR is way too encompassing of features.

tesfaldet committed:
Added agent budget functionality for launching batches of agents until budget reached. Added some comments. Retrieving results from task function via its agent so that it can be visible to the sweeper and launcher for inspecting JobReturn status and return value.

captain-pool commented on example/my_app.py:
```diff
@@ -0,0 +1,37 @@
+    Minimum is 0.0 at:
+    lr = 0.12, dropout=0.33, db=mnist, batch_size=4
+
+    to execute code --> python my_app.py --multirun +sweeper=wandb
```
I think we would prefer overriding default hydra's sweeper instead of doing something like this. For example:

we would be more interested in having a plugin which can be invoked as `python my_app.py --multirun hydra/sweeper=wandb`

tesfaldet replied:
I believe it should still be able to be executed that way no problem. Let me test...

tesfaldet replied:
The above example was to showcase that you could provide multiple overrides in the [style mentioned in the Hydra docs](https://hydra.cc/docs/patterns/configuring_experiments/). Specifically, you could do something like `python my_app.py --multirun +experiment=[mnist, wandb_sweep]` where the `override /hydra/sweeper: wandb` line would be in `wandb_sweep.yaml` for example. Anyways, I'll test out `python my_app.py --multirun hydra/sweeper=wandb` and report back; I'm pretty sure it should work.

tesfaldet replied:
[It works](https://wandb.ai/tesfaldet/hydra_plugin_test_project/sweeps/fwd2rxks?workspace=user-tesfaldet) :)

I commented out 
```yaml
defaults:
   - override /hydra/sweeper: wandb
```
from `wandb.yaml` and executed `python my_app.py --multirun hydra/sweeper=wandb +sweeper=wandb` just to see if it'd still use the wandb sweep config defined in `conf/sweeper/wandb.yaml` and it did. If you prefer me to change the example to follow this then I surely can no problem, I just like the example explicitly showing that you can merge override configs with the default config which includes overriding the sweeper.

The code that verifies the type of override is here:
https://github.com/captain-pool/hydra-wandb-sweeper/blob/cbda3c61e9c55d48fb338aec63e4d3e6c275ac91/hydra_plugins/hydra_wandb_sweeper/wandb_sweeper.py#L137-L146

captain-pool replied:
Amazing! Thanks again for your wonderful contribution! I'll look into it next week and get back to you shortly. Just as a heads-up, we will be shifting this repository to [github.com/wandb](https://github.com/wandb) and will be making a PR to hydra once everyone is happy about the basic functionalities. So please feel free to fix anything I might have missed. (for example the 2-space yaml indentation problem, you can fix that by modifying the `offset` parameter in `yamlfmt` hook inside [`.pre-commit-config.yaml`](https://github.com/captain-pool/hydra-wandb-sweeper/blob/main/.pre-commit-config.yaml) file. You can find more about it [here](https://github.com/jumanjihouse/pre-commit-hook-yamlfmt/blob/master/README.md).)

tesfaldet replied:
Awesome! I can make that change in a separate PR :) I'm assuming the repo transfer will include all the commit history?

captain-pool replied:
Regarding repo transfer. I'll try my best to transfer all the commit history properly during the PR and the transfer. Also, if something weird happens during the transfer, I'll make sure to highlight your extensive contributions manually in the README.

tesfaldet replied:
[Fortunately repo transfer while preserving history should be painless :)](https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository#whats-transferred-with-a-repository)

tesfaldet commented:
Switching PR back to draft since I just found a weird bug for saving the wandb sweep config yaml. The culprit is here:
https://github.com/captain-pool/hydra-wandb-sweeper/blob/cbda3c61e9c55d48fb338aec63e4d3e6c275ac91/hydra_plugins/hydra_wandb_sweeper/wandb_sweeper.py#L90-L92

What's happening is that it's creating a folder `../../hydra-.wandb-sweeper/example/multirun/2022-02-28/13-24-24/wandb/sweep-fwd2rxks/` instead of `multirun/2022-02-28/13-24-24/.wandb/sweep-fwd2rxks/`. I need to be more rigorous in my `config_filename` check... The problem is when the working directory contains `wandb` in a valid manner that has nothing to do with the actual `wandb` save directory where `config_filename` is meant to save to. This happens because the check `if ".wandb" not in config_filename and "wandb" in config_filename` is being performed on an absolute filepath, which can include other usages of the word `wandb` in the path, for example: `/home/mila/m/mattie.tesfaldet/Libraries/hydra-wandb-sweeper/example/multirun/2022-02-28/14-16-03/wandb/sweep-0frl3j64/config-0pq8snsk.yaml`. The intention is to replace `/wandb/` in the path with `/.wandb/` and not any other instance of it. I'm gonna implement a fix.

EDIT: fixed

tesfaldet committed:
Returning results and statuses from task function to launcher. Improved task function error handling and propagation to launcher. Fixed bug with renaming wandb dir to .wandb when config_filepath has text 'wandb' anywhere that doesn't relate to a directory.

tesfaldet committed:
Moved main sweeper code to _impl.py so that it doesn't get accidentally loaded along with all its monkeypatching when the user doesn't want to use the sweeper. Upon running a task function, hydra loads all available plugins, even those that won't be used. Improved code typing. Added implicit start to yamlfmt pre-commit since it clashed with hydra's # @package _global_ indicator.

tesfaldet commented:
Heads up that I'm still very much working on this PR! I still need to deal with the TODOs above, specifically the pre-emption stuff which I feel is VERY important.

I'm trying to run sweeps for my research project and the annoying rate limits are causing agents to halt my hydra task functions / slurm jobs, which are timing out (I'm currently in an email thread with Leslie's who's helping me with this). I haven't thoroughly looked into `wandb_agent.py` to understand the halting process but I have a feeling that it's straight up exiting the thread that's running the task function and re-running it. Since I'm using submitit (via the [hydra submitit launcher plugin](https://hydra.cc/docs/plugins/submitit_launcher/)) to launch agents distributed across a SLURM cluster (each SLURM job running exactly 1 agent which is running its task function exactly once), I am relying on submitit's [useful checkpointing functionality](https://github.com/facebookincubator/submitit/blob/main/docs/checkpointing.md) to catch any `SIGTERM` or `USR1` signals which the underlying `sbatch` that submitit generated sends to the unpickled submitit job upon its pre-emption and/or timeout (you can find the `checkpoint` function [here](https://github.com/facebookresearch/hydra/blob/9cc91b995a0f2bf52d57a3c878721e42c2056995/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/submitit_launcher.py#L79-L84)). Unfortunately, since wandb agents wrap the task function in its own multithreading context, it means that the hydra launcher (in this case, submitit) isn't aware of what's going on (although I have made some [improvements](https://github.com/captain-pool/hydra-wandb-sweeper/pull/5/commits/45cf16a19a1b8db43f9f59a7f35eb6cbfe297dee) on that front). So if an agent decides to halt a task function due to rate limiting, I have a feeling there's no easy way to resume. `wandb.mark_preemption()` wouldn't apply here since that's meant for SLURM-level halting, not for `wandb.agent`-level halting. Anyways, I need to continue looking into this for the sake of my research hahaha...

I really wish the rate limits weren't so stringent for sweeps. It's odd because that's not the case for non-sweep runs. You can run waaaaaay more runs that aren't part of a sweep and log them far more frequently and you'd have no problem. As soon as you're using agents and a sweep however, you're slammed with limits.

captain-pool commented:
Amazing! I was actually planning to provision myself a cluster from W&B to test out these things. 😄  But looks like you've got it covered (I might however take an AWS Cluster to check the Ray AWS Cluster integration).

Regarding halting, I think `wandb.mark_preempting()` can deal with this. If you look into the documentation, it says `If a run exits with no status, W&B will mark the run preempted 5 minutes after the run's final heartbeat, then add it to the sweep run queue.` Hence even if it's slurm level halting, you shouldn't have any issue regarding it.

Thanks again for all your contributions 🙂  I'll go through the PR once my leave gets over and I join back office next week.

tesfaldet commented:
Ah that's useful of wandb.

So I looked into `wandb_agent.py` and my `debug-internal.log` as well as my other log files and found out that rate limiting doesn't halt the process 😅 It just doesn't get logged and the task function continues to chug along. This is very good. That means that any errors within the task function will come from non-wandb-related code, and any pre-emption/timeouts/halting or whatever will come from SLURM (or whatever cluster controller), which is how it should be. As you've probably noticed in one of my commits, I've added mechanisms to allow any Hydra launcher to inspect the status and returns of a wandb-wrapped task function. So I could piggy-back off of that and test out `wandb.mark_preempting()` in the way I mentioned previously. Specifically by providing my own `MyBaseSubmititLauncher` class that is a subclass of `BaseSubmititLauncher` with an override on the `checkpoint` function that will have `wandb.mark_preempting()` in it. [This is something I've become quite familiar with over the past couple of weeks](https://github.com/facebookresearch/hydra/issues/2042#event-6141048658).

Additionally, I'm going to add checks on returned batches of launched jobs to inspect and count the number of failed jobs in order to cancel launching any new batches of jobs. This is inspired by the [nevergrad launcher plugin](https://github.com/facebookresearch/hydra/blob/9cc91b995a0f2bf52d57a3c878721e42c2056995/plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/_impl.py#L166-L188) and it should help avoid the launcher from repeatedly launching jobs when a whole lot of previous ones were failing. I'm also going to add checks on the sweep controller's status to verify that it's still alive to prevent the launcher from launching jobs when the sweeper dies mid-run (e.g., killing a sweep via the web interface). EDIT: implemented improved error handling related to this.

tesfaldet committed:
Improved error handling for agent failures such as sweep not existing anymore, job failures unrelated to the agent, and run failures within an agent.

tesfaldet commented:
Update: still working on this. I've been testing the pre-emption mechanism with the submitit launcher over the past several days and it's been difficult because of the all the multithreading going on between `wandb.agent`, `launcher.launch`, `wandb.init`, etc. Lots of making sure things are cleaned up during pre-emption, catching system exits and propagating them upwards to the launcher, etc. However, I think I've got something....

tesfaldet committed:
Revamped example training script to showcase new pre-emption capabilities via the submitit launcher plugin. Note, not tested with Joblib, Ray, or RQ launcher plugins.

tesfaldet commented:
![Screen Shot 2022-03-05 at 15 29 28](https://user-images.githubusercontent.com/5014902/156899056-e590a995-67d0-44ce-b228-dc53211ed441.png)
![Screen Shot 2022-03-05 at 15 36 53](https://user-images.githubusercontent.com/5014902/156899253-c5fe5b6f-0139-497f-9b2f-c6ea09938cb4.png)

I managed to get Hyperband early termination to work with this plugin :) At least for the most part. It works with both the basic launcher and the submitit launcher, but for the latter there's an odd bug I'm experiencing when I have multiple agents running in parallel. Something I'm looking into rn.

Furthermore, with my testing of the Hyperband early termination feature, I also managed to iron out some bugs pertaining to stopping sweeps, killing sweeps, and stopping individual runs manually through the wandb interface. It all works fine so far, even with multiple parallel agents. The only problem seems to be when the stop request comes from hyperband with multiple parallel agents.

EDIT: issue seemed to be related to threads locking resources needed by other threads during the `_poll_exit_response` loop in `_on_finish()` in `wandb_run.py`. It would cause `os._exit(-1)` to be called, which would terminate all parent threads, including the agent and `WandbTask` task function, which would cause either the submitit plugin's `map_array` to yell since no output pickle was produced (if using submitit) or the entire hydra process to terminate (if using the basic launcher). This was on version 0.12.1 and after inspecting 0.12.11's release code, I noticed some changes in `_on_finish()` that may fix the issue, so I updated to this version and voila, problem gone :) Perhaps in the future it might be worth using [wandb service](https://github.com/wandb/client/blob/master/docs/dev/wandb-service-user.md) instead.

tesfaldet committed:
Support for Hyperband early termination. Graceful response towards remote killing/stopping of sweeps and runs.

captain-pool commented on .pre-commit-config.yaml:
```diff
@@ -1,43 +1,48 @@
+          - parsable
+          - --strict
+          - --config-data
+          - '{extends: default, rules: {document-start: disable}}'
```
If you are disabling `document-start` then please remove `exclude: example/conf/(logging)` from the `yamlfmt` hook.

captain-pool commented on README.md:
````diff
@@ -11,29 +11,15 @@ $ pip install git+https://github.com/captain-pool/hydra-wandb-sweeper.git
 ```python
 @hydra.main(config_path=..., config_name=...)
-def main(cfg: DictConfig, run_id: str): # notice the extra run_id parameter
+def main(cfg: DictConfig):
````
The design decision that went behind adding the `run_id` parameter was to give the user explicit control over adding more parameters to `wandb.init()` as they normally would do, without creating a new run. This was specifically targetted to people who wanted to port their W&B projects to hydra with minimal changes in the py file. How doe it work in those cases?

tesfaldet replied:
There were a couple of reasons why I changed this:
1. Having a wandb.init just for retrieving a run id and config and passing it through to the task function was unnecessary overhead since you could just run the task function within the wandb.init context, as I showed in the example.
2. By removing that argument above, it keeps things in parity with what Hydra expects. Even if you were to make it an optional parameter, it avoids any weird bugs that may pop up in the future due to whatever function introspection Hydra does. Basically, I'm just trying to avoid any unnecessary problems.

In light of 1 and 2, the reason I say it's unnecessary to explicitly pass the run id is because since you're within a [wandb run context already](https://github.com/captain-pool/hydra-wandb-sweeper/blob/e791aec734530bc9b611f938ca6a5962d67eb6e6/hydra_plugins/hydra_wandb_sweeper/_impl.py#L619), you can just retrieve the run id by calling `wandb.run.id`.

captain-pool replied:
The first `wandb.init` (inside `wandb_task`) was added to only send Hydra Configs to the run spawned by the agent. The `run_id` was basically a byproduct which allowed continuing the started run, inside main, without needing a lot of change in the user-cdoe. But, I guess, you're correct. It was adding a lot of overhead. And since we are in a wandb context, an user can also open another `wandb.init()` context inside main and pass any extra features they want to add. Since they are not allowed to create new runs inside an agent, an user should be able to continue with the same run with new parameters that they want to specify.

tesfaldet replied:
I don't believe running another `wandb.init()` is necessary within main since you can directly access the run by calling `wandb.run`. From there you can use any of the [setters](https://github.com/wandb/client/blob/master/wandb/sdk/wandb_run.py) within the run object. The user shouldn't be running `wandb.init()` again since it's just additional overhead. What they could do instead, if they want to port over their code, is to check if a currently active run is available (`if wandb.run is not None`), which would imply that they're within the run that was set from WandbTask, and from there retrieve the run and use its setters to modify things like tags, notes, etc. If they're not within a currently active run, then they can `wandb.init()`.

captain-pool commented on example/conf/sweeper/wandb.yaml:
```diff
@@ -0,0 +1,69 @@
+    # parameters are described using dot notation but will be accessible within
+    # the task function in typical OmegaConf fashion,
+    # e.g., model_param1 = cfg.model.param1
+    params:
+      # categorical, one or the other
+      experiment.db:
+        - mnist
+        - cifar
+
+      # a log-distributed positive scalar, evolving by factors of 2 on average
+      experiment.lr:
+        distribution: q_log_uniform
+        min: 0.02
+        max: 1
+        q: 2.0
+
+      # a linearly-distributed scalar between 0 and 1
+      model.dropout:
+        distribution: uniform
+        min: 0.0
+        max: 1.0
+
+      # a linearly-distributed integer scalar between 4 to 16
+      experiment.batch_size:
+        distribution: int_uniform
+        min: 4
+        max: 16
```
This is amazing! Thanks for doing this. I think having configurations inside a yaml allows the user to use sweeps without writing a long list of command line arguments.

tesfaldet replied:
That's what Hydra's meant for! All the other sweepers follow something similar. Furthermore, they're standardizing this approach further in v.1.2 https://github.com/facebookresearch/hydra/issues/1890

tesfaldet replied:
Just a heads up that grid-type hyperparam sweeps still needs to be supported, hence the PR still being in draft state. I don't feel comfortable putting it in review state until I've got that sorted out since that's still an essential part of the wandb hyperparam sweep mechanism, lest we want an incomplete sweeper :)

captain-pool commented on example/hydra_plugins/launchers/my_submitit_launcher.py:
```diff
@@ -0,0 +1,65 @@
+# submitit won't be able to find the my_submitit_launcher module. A way around this is to make a nested folder,
+# 'launchers', with an __init__.py and place the module there. Then from train.py we register
+# hydra_plugins.launchers by value.
+class MyBaseSubmititLauncher(BaseSubmititLauncher):
```
Please use relevant class names,  not something ilke this.

tesfaldet replied:
Got any ideas? It's meant to be an example, not a part of the core sweeper code. It's fairly common (and not unpythonic) to use "My" as part of the naming of an example class, function, variable, etc. https://www.python.org/dev/peps/pep-0008/ even has some examples here. Specifically, for this example the naming was inspired from this [GH issue](https://github.com/facebookresearch/hydra/issues/2042#issuecomment-1045333597) I made for Hydra a couple of weeks ago.

captain-pool replied:
Agreed. It's very common to use `My`. I got confused by the comments and thought this is required for making W&B work with submitit. If it is not a core requirement, we can keep it this way. If it is a core requirement for the submitit integration, we can rename it to something like `WandbSubmititLauncher`.

tesfaldet replied:
It's not a core requirement for the submitit integration since that's all up to the user. Basically, I tried making the sweeper as launcher agnostic as possible, despite not testing for the other launchers. The way I did that was by designing the sweeper implementation around the mindset of "What happens if a SystemExit is called by the task function? What if it comes from the agent? What if it comes from Hydra? What if it's a os._exit where it's not an exception and everything just blows up ungracefully (as wandb does in `_on_finish()`)? What happens if signals are being sent to this process? How would I deal with this if wandb already does some signal handling itself?" With this mindset, it's fairly unspecific to SLURM and submitit but it just happens to cover those bases since SLURM uses signalling for pre-emption. It's why I provided an extensive example to truly test the limits of this plugin with some sort of pre-emption or timeout mechanism, otherwise users will have no clue since it's fairly complex (partially because of wandb limitations and hydra limitations). If the other cluster-based launchers do the same then great! If not, well, it's up to the user to figure it out and it shouldn't be up to this sweeper. Again, I kept it as launcher agnostic as possible because it's not the sweeper's job to figure out what launcher it's playing with. I know I mention "submitit" in the comments of `_init.py` here and there but really it applies to any cluster-based launcher.

I think testing out the other launchers should be a separate PR.

captain-pool commented on example/hydra_plugins/launchers/my_submitit_launcher.py:
```diff
@@ -0,0 +1,65 @@
+    def __init__(self, **params: Any) -> None:
+        super().__init__(**params)
+
+    def checkpoint(self, *args: Any, **kwargs: Any) -> Any:
```
Not much aware of submitit functionalities. Does this checkpoint only get called when you want a cluster node preempties? That's because, I can see in some of the lines, that you are enforcing the run to restart when this function gets called. Might be problematic if this gets called in different circumstances (i.e., non-preempting cases)

tesfaldet replied:
More specifically, it gets called when `SIGTERM` and/or `USR1` and/or `SIGCONT` are sent to the parent process and _only_ in a SLURM environment (it doesn't get used when using submitit_local since that's a local environment). It does not get accidentally called by wandb, if that's what you're worried about :) 

This can either happen from job timeouts when the process hasn't finished yet the allotted time has finished: in this case, SLURM signals `USR1` 120 seconds (#seconds can be changed) before the job times out (read about `USR1` coming from `sbatch` [here](https://slurm.schedmd.com/sbatch.html#OPT_signal)), then `SIGCONT`, then `SIGTERM`, then finally `SIGKILL` when the job has timed out. `checkpoint` gets called on `USR1` in this case and does the necessary cleanup and job resubmission before exiting the process itself (before `SIGKILL` arrives from SLURM). In the pre-emption case, there is no `USR1` and instead `SIGCONT` (and immediately afterwards `SIGTERM`) arrive, which gets caught by submitit and executes `checkpoint`, which gives it time to clean up, resubmit, and exit before `SIGKILL` arrives. You can read more about pre-emption under "Grace Time" [here](https://slurm.schedmd.com/preempt.html).

Just to make it clear, this does not explicitly restart the run since that's impossible on wandb's end at the moment. Wandb sweeps don't allow you to restart specific runs, which would've been really nice. Once this checkpoint function has marked the run as pre-emptible and finishes with a non-zero exit code, there's a chance that any of the other sweep agents currently running could pick up the requeued run (i.e., the same run parameters) before the checkpoint function has successfully resubmitted the task function (in this case, `__call__` in `DummyTraining`) and that task function has begun a sweep agent to pull runs from the sweep controller. Basically, `mark_preempting` just puts the run params back up on a global queue that ALL agents can pull from. What would be really nice would be if:
1. You can restrict wandb agents from pulling requeued runs from the wandb sweep controller. That way, when you specify `mark_preempting` during pre-emption, you can configure the next agent that will be a part of the resubmitted task_function to be able to pull the requeued run from the controller.
2. You can specify the run ID of the run to be initialized and if that run ID was an ID used in a pre-empted run, it'll pull the sweep params that were associated with that ID. That way, when you resubmit your task function during pre-emption, once it spins up its agent, it'll know exactly which run to continue from.

captain-pool commented on example/hydra_plugins/launchers/my_submitit_launcher.py:
```diff
@@ -0,0 +1,65 @@
+        return super().checkpoint(*args, **kwargs)
+
+
+class MyLocalLauncher(MyBaseSubmititLauncher):
+    _EXECUTOR = "local"
+
+
+class MySlurmLauncher(MyBaseSubmititLauncher):
+    _EXECUTOR = "slurm"
```
Again, better class names.

captain-pool commented on example/my_app.py:
```diff
@@ -0,0 +1,106 @@
+            f"= {best:.3f}"
+        )
+
+        if cfg.error:
+            raise RuntimeError("cfg.error is True")
```
This kind of tasks should be in unittests, not in examples.

tesfaldet replied:
Agreed. Honestly I was inspired by https://github.com/facebookresearch/hydra/blob/796fdf1c54eabdcd5121d3a9c8ab6af70ee41763/plugins/hydra_nevergrad_sweeper/example/my_app.py#L24-L25 and wanted to quickly test out runtime errors during a sweep. I'll remove it.

captain-pool commented on example/my_app.py:
```diff
@@ -0,0 +1,106 @@
+
+if __name__ == "__main__":
+    dummy_training = DummyTraining()
+    app = hydra.main(config_path="conf", config_name="config")(dummy_training.__call__)
```
If this is a SLURM specific example, please mention that. I'd also request you to add a minimal example with local multiprocessing (with ray or joblib). That way, people who don't want to use SLURM can have something to refer to.

An example that can be invoked in the following way.
```bash
$ python3 main.py hydra/sweeper=wandb hydra/launcher=ray
$ python3 main.py hydra/sweeper=wandb hydra/launcher=joblib
```

tesfaldet replied:
I mention it here https://github.com/captain-pool/hydra-wandb-sweeper/blob/e791aec734530bc9b611f938ca6a5962d67eb6e6/example/my_app.py#L19-L37

Submitit also supports local multiprocessing by switching to: `- override /hydra/launcher: submitit_local`
Documentation available [here](https://github.com/facebookincubator/submitit/blob/main/docs/examples.md) and [here](https://hydra.cc/docs/plugins/submitit_launcher/).

Explicit support for the other launchers, such as Ray and Joblib, go beyond the scope of this PR. If people want to do local multiprocessing they can use the submitit launcher but if they want to try out Ray or Joblib, that'll have to wait (I don't have the time since I made this PR for the sake of helping out my research and I need to get back to my research).

captain-pool replied:
Cool! Once this PR gets merged, I'll take look into ray / joblib. :) I had no idea about `submitit_local`. Thanks for the heads up.

tesfaldet replied:
Just a heads up that this sweeper should be launcher agnostic as much as possible! So if it doesn't work for the other launchers lmk and we can figure out a way to future proof this. It would be a nightmare if down the road the other launchers get modified and suddenly this sweeper needs to be modified to reflect that change.

captain-pool commented on hydra_plugins/hydra_wandb_sweeper/config.py:
```diff
@@ -1,30 +1,95 @@
+    # (float) Quantization step size for quantized hyperparameters.
+    q: Optional[float] = 1
+
+    # Specify how values will be distributed
+    # Supported distributions and required parameters
+    # "constant": ["value"],
+    # "categorical": ["values"],
+    # "int_uniform": ["min", "max"],
+    # "uniform": ["min", "max"],
+    # "q_uniform": ["min", "max", "q"],
+    # "log_uniform": ["min", "max"],
+    # "q_log_uniform": ["min", "max", "q"],
+    # "inv_log_uniform": ["min", "max"],
+    # "normal": ["mu", "sigma"],
+    # "q_normal": ["mu", "sigma", "q"],
+    # "log_normal": ["mu", "sigma"],
+    # "q_log_normal": ["mu", "sigma", "q"],
```
If this section of code is not needed, please remove it instead of commenting out the section.

tesfaldet replied:
It's not code, it's just comments to inform the support distributions and their associated required parameters. It's meant for those who decide to look into this code in the future.

captain-pool commented on hydra_plugins/hydra_wandb_sweeper/config.py:
```diff
@@ -1,30 +1,95 @@
-
-@dataclasses.dataclass
+    name: str = "hydra_wandb_sweep"
+    method: str = "bayes"
```
Can we make this a required parameter instead of an optional one? Don't think setting it to bayesian makes sense.

tesfaldet replied:
Oof, so this one is tricky to explain. There is a very odd Hydra-specific situation where it's necessary for this to have default values. Specifically, because of submitit pre-emption via the Hydra submitit plugin.

I agree it should be a required parameter and since I want this to be as launcher agnostic as possible, I shouldn't be making a decision such as above because of peculiarities from the launcher. I'll find a workaround.

tesfaldet replied:
I forgot that OmegaConf has the `MISSING` signifier for required parameters that I could test out. It could perhaps workaround the issue I briefly mentioned. I'm taking a look at the [Optuna sweeper config as an example](https://github.com/facebookresearch/hydra/blob/796fdf1c54eabdcd5121d3a9c8ab6af70ee41763/plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/config.py#L140).

captain-pool commented on hydra_plugins/hydra_wandb_sweeper/wandb_sweeper.py:
```diff
@@ -1,174 +1,31 @@
 
-def get_wandb_parameter(override):
+    def __init__(self, wandb_sweep_config: WandbConfig, params: Optional[DictConfig]):
+        from hydra_plugins.hydra_wandb_sweeper._impl import WandbSweeperImpl
```
Can you please use `importlib.import_module` instead of this?

tesfaldet replied:
Hahaha yeah I should've done that :) I was semi-blindly following the nevergrad sweeper implementation https://github.com/facebookresearch/hydra/blob/796fdf1c54eabdcd5121d3a9c8ab6af70ee41763/plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/nevergrad_sweeper.py#L16

I agree it's better to use `importlib.import_module` when programatically importing modules.

captain-pool reviewed (changes requested):
All in all the PR looks amazing! Apologies for getting back this late. Thank you for your wonderful contribution 🎉 . There are some minor changes that I requested. Apart from that, it looks amazing. Let me know if you need any help. I'll be happy to help you out in writing an unittest for this.

tesfaldet commented:
> All in all the PR looks amazing! Apologies for getting back this late. Thank you for your wonderful contribution 🎉 . There are some minor changes that I requested. Apart from that, it looks amazing. Let me know if you need any help. I'll be happy to help you out in writing an unittest for this.

Hey no worries! Just a heads up that the PR is still a draft though! I'm still actively making changes as per the TODO above. However, I will comment on the requests you've made pertaining to the TODO items I have completed so far. Once I mark the PR as ready to review then that's when it should _actually_ be reviewed for merge.

tesfaldet committed:
More comments for clarity. Tweaking example so that it's easier to test out wandb alert during preemption. Slight code tweaks elswhere. Small pre-commit modifications.

tesfaldet committed:
Grid search method now supported for both config-style parameterization and override-style. Validation of metric arguments, early termination args, and search args.

tesfaldet commented:
Update: Grid search method now supported. It works for both config-style parameterization and command-line override-style parameterization. Here are some examples...

Examples of multiple values:
```yaml
parameter_name:
  values:
    - 8
    - 9
```

```yaml
parameter_name:
  values: [8, 9]
```

```yaml
parameter_name:
  - 8
  - 9
```

```yaml
parameter_name: [8, 9]
```

Examples of single values:
```yaml
parameter_name:
  value: 8
```

```yaml
parameter_name: 8
```

Gotcha cases that are covered:
1. Providing a single value for `values`
```yaml
parameter_name:
  values:
    - 8
```
gets converted internally to `values`.

2. Providing multiple values for `value`
```yaml
parameter_name:
  value:
    - 8
    - 9
```
gets converted internally to `values`.

captain-pool commented on example/conf/sweeper/wandb.yaml:
```diff
@@ -0,0 +1,113 @@
+    # See https://docs.wandb.ai/guides/sweeps/configuration for types of configs
+    wandb_sweep_config:
+      # (Required) The name of the sweep, displayed in the W&B UI.
+      name: hydra_plugin_test
+
+      # (Required) Specify the search strategy.
+      method: bayes
+
+      # (Optional) Specify the metric to optimize (only used by certain search
+      # strategies and stopping criteria). Default is {}.
+      metric:
+        name: out
+        goal: minimize
+
+        # (Optional) Goal value for the metric you're optimizing. When any run
+        # in the sweep achieves that target value, the sweep's state will be set
+        # to finished. This means all agents with active runs will finish those
+        # jobs, but no new runs will be launched in the sweep.
+        # target: 0.0
+
+      # (Optional) Number of agents to launch in a batch until budget is
+      # reached. Use a multi-processing launcher to have these agents execute
+      # simultaneously, otherwise it's a useless option since they'll execute in
+      # sequence. Default is 1.
+      num_agents: 2
+
+      # (Optional) Total number of agents to launch (must be >= num_agents)
+      # Default is 1.
+      budget: 4
```
Some of the parameters here looks pretty similar to what an user would have in a wandb `sweep.yaml` file. Would it be possible for you to group the `sweep.yaml` parameters together and wrap them within a comment block like:
```
######### Beginning sweep.yaml ############
method:
metric:
. . .
######## End sweep.yaml ################

<other wandb_sweeper_config specific params>
````

This would allow an user to easily port a wandb sweeper project into hydra without worrying about learning new parameters for this plugin.

Additionally, it would also be cool to add support to loading wandb's sweep.yaml file inside in here so that an user doesn't need to copy contents from a previously existing sweep.yaml in here. But that's low priority.

tesfaldet replied:
Hmmm, this is a tough one because Hydra is aiming for a [standardized approach to sweep configurations](https://github.com/facebookresearch/hydra/issues/1890). I wouldn't really want to confuse hydra's templating engine buuut there could be a sort of optional transpiling process to convert from wandb's format to Hydra's format internally before the start of a sweep. I think that should be a separate PR however since that could be quite extensive.

captain-pool commented on hydra_plugins/hydra_wandb_sweeper/_impl.py:
```diff
@@ -0,0 +1,856 @@
+                        agent._setup()
+                    except Exception as e:
+                        agent_id = None
+                        agent_error = "Sweep has been killed."
```
I think there can be other reasons as well, not just killed. For example, IIRC, if an user stops / pauses a sweep, and tries to create an agent without resuming the sweep, the agent throws an error.

tesfaldet replied:
Ah, fair point. So perhaps a better message is "Sweep is not available"?

captain-pool replied:
How about extracting the message from the exception (using `str(e)`) and using that instead?

tesfaldet replied:
Hmmm, that will make it difficult to figure out if a sweep is available or not when the next batch of agents are to be launched, [since that error message is what's used to figure out if a sweep is available](https://github.com/captain-pool/hydra-wandb-sweeper/blob/c7211452e7b1a50924aa768cb315206f1eb29447/hydra_plugins/hydra_wandb_sweeper/_impl.py#L654-L664). Don't get me wrong, I agree with what you're saying, I'm just trying to figure out another way of detecting whether the sweep is available or not, given the sweep id, before launching the next batch of agents.

tesfaldet replied:
@captain-pool do you happen to know of a programmatic way of querying whether a sweep is available or not given a sweep id?

captain-pool replied:
Apologies for the delay. I checked in with the WandB sweeps team. As of now, there’s currently no way of doing that. Although you might find a `state` property in the documentation for sweeps, which for some reason is non-existent in the code. I will raise an internal ticket to notify the docs team, and raise a feature request with the Sweeps team.

tesfaldet replied:
Ah darn. Thanks for raising the feature request :)

captain-pool reviewed (approved):
To me it looks great. I'm adding my colleague @dmitryduev to the conversation to look over the codebase once again and check if I missed anything.

captain-pool commented:
CC: @dannygoldstein

captain-pool commented:
Eitherway, I’m merging this pull request, Thanks again for your wonderful contribution. I’ll notify you when I create a pull request with Hydra.
